zkSNACKs/WalletWasabi
# Make software update flow better

This PR implements an `UpdateManager.cs` class which makes updating after a new release easier.

**Old flow:**
1. In case of new version, Wasabi indicates to "Update".
2. Click on "Update" button.
3. Get redirected to Wasabi website.
4. Find your version to download based on OS.
5. Install manualy.

**New flow** is somewhat simpler:
1. In case of new version, Wasabi downloads the version you need in the background, then indicates "Close and Update".
2. Click on "Close and Update".
3. Upon closing, the installer will start automatically.

## How to test
1. Find `WalletWasabi/Helpers/Constants.cs`.
2. Change `ClientVersion` to: `new(2, 0, 0, 1)` or any other version thats smaller then current.
3. Start Wasabi and go through the steps above.
4. You can also double check logs and see if `Roaming\WalletWasabi\Client\Downloads` folder gets deleted after succesful install.

Tested on Windows. ✅ 
Tested on M1 MacOS. ✅ 
Tested on non-M1 MacOS. ✅ 
Does not support other linuxes currently, sorry Max.

## TODOs

- [x] Use .tmp file format until download finished.
- [x] Add "Update on close" to settings (config file, default: `true`).
- [x] Check if works on non-M1 Mac.
- [ ] ~Use tor for downloading installer as well.~ _Not prio, will be added later, a PR will be opened for this not to be forgotten._
- [ ] ~Restart Wasabi after updating(?)~ _Bumped, best way IMO is to add a "Start Wasabi" checkable option at the end of the installer._
- [ ] ~Download `.asc` and verify GPG as well.~ _In later PR so this PR don't get overwhelmed_
- [ ] ~Terminate Tor on close if "Update on close" is `true`.~ _Later in different PR, need to add `ShutDownTor` method to TorProcessManager_
